### PR TITLE
Add incident_io_user Data Source

### PR DESCRIPTION
**Pull request by rorymalcolm**
- Adds a `data` source, `incident_io_user`

- This lets users fetch information about incident.io users in the Terraform provider

- Have explicitly decided _not_ to add acceptance tests here - data sources are obviously materially different from our resource tests!

### CODE REVIEW

File: internal/apischema/openapi3.json [hunk 1]
```diff
@@ -783,6 +783,7 @@
                 "github_pull_request",
                 "gitlab_merge_request",
                 "sentry_issue",
+                "jira_issue",
```
Comment (henryjcee) [inline internal/apischema/openapi3.json:786]: is this intentional

File: internal/provider/incident_user_data_source.go [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+			return
+		}
+		user = &result.JSON200.User
+	} else if !data.Email.IsNull() {
+		result, err := i.client.UsersV2ListWithResponse(ctx, &client.UsersV2ListParams{
+			Email: data.Email.ValueStringPointer(),
+		})
+		if err == nil && result.StatusCode() >= 400 {
+			err = fmt.Errorf(string(result.Body))
+		}
```
Comment (henryjcee) [inline internal/provider/incident_user_data_source.go:96]: In what circumstances do we get a nil error and 4++ response?

**Review by henryjcee — APPROVED** [review]
One question, otherwise this looks sensible.

File: internal/provider/incident_user_data_source.go (hunk 2 above)
Comment (rorymalcolm) [reply internal/provider/incident_user_data_source.go:96]: Same as what we do elsewhere in the provider!

File: internal/provider/incident_user_data_source.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+func NewIncidentUserDataSource() datasource.DataSource {
+	return &IncidentUserDataSource{}
+}
+
+type IncidentUserDataSource struct {
+	client *client.ClientWithResponses
+}
+
+type IncidentUserDataSourceModel struct {
+	BaseRole    *RBACRole    `tfsdk:"base_role" json:"base_role"`
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:29]: Do we care about roles here? If not, should we just slim this down to what we actually need to make our life simpler?

File: internal/provider/incident_user_data_source.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	CustomRoles []*RBACRole  `tfsdk:"custom_roles" json:"custom_roles"`
+	Email       types.String `tfsdk:"email" json:"email"`
+	ID          types.String `tfsdk:"id" json:"id"`
+	Name        types.String `tfsdk:"name" json:"name"`
+	Role        types.String `tfsdk:"role" json:"role"`
+	SlackUserId types.String `tfsdk:"slack_user_id" json:"slack_user_id"`
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:35]:
```suggestion
	SlackUserID types.String `tfsdk:"slack_user_id" json:"slack_user_id"`
```
would be more go-y

File: internal/provider/incident_user_data_source.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+type IncidentUserRequest struct {
+	ID types.String `tfsdk:"id"`
+}
+
+type RBACRole struct {
+	Description types.String `tfsdk:"description"`
+	Id          types.String `tfsdk:"id"`
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:44]:
```suggestion
	ID          types.String `tfsdk:"id"`
```
the inconsistency is killing me

File: internal/provider/incident_user_data_source.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	i.client = client
+}
+
+func (i *IncidentUserDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_user"
+}
+
+func (i *IncidentUserDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data IncidentUserDataSourceModel
+	req.Config.Get(ctx, &data)
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:73]: Do we want to do the `resp.Diagnostics.Append` dance here?

File: internal/provider/incident_user_data_source.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+		if err == nil && result.StatusCode() >= 400 {
+			err = fmt.Errorf(string(result.Body))
+		}
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read user, got error: %s", err))
+			return
+		}
+		user = &result.JSON200.User
+	} else if !data.Email.IsNull() {
+		result, err := i.client.UsersV2ListWithResponse(ctx, &client.UsersV2ListParams{
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:91]: I quite dislike this vs. having a lookup endpoint where you can provide an email, but I think we've had this for ages so I'll code

File: internal/provider/incident_user_data_source.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+			Email: data.Email.ValueStringPointer(),
+		})
+		if err == nil && result.StatusCode() >= 400 {
+			err = fmt.Errorf(string(result.Body))
+		}
+		if err != nil {
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read user, got error: %s", err))
+			return
+		}
+		if len(result.JSON200.Users) == 0 {
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:101]: I think we might want to log.Debug if we get more than one user?

File: internal/provider/incident_user_data_source.go [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+			resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read user, got error: %s", "User not found"))
+			return
+		}
+		user = &result.JSON200.Users[0]
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:121]: same here, should never happen, but lets scream if we have >1

File: internal/provider/incident_user_data_source.go (hunk 7 above)
Comment (rorymalcolm) [reply internal/provider/incident_user_data_source.go:91]: I just added it today lol

Comment (paprikati) [reply internal/provider/incident_user_data_source.go:91]: hmm. I don't love this in our public API vs. an explicit lookup endpoint. One for #api-design? Not a hill I'll die on

File: internal/provider/incident_user_data_source.go (hunk 3 above)
Comment (rorymalcolm) [reply internal/provider/incident_user_data_source.go:29]: I think it's much easier if this is a faithful representation of what the API has in it

Comment (paprikati) [reply internal/provider/incident_user_data_source.go:29]: But as soon as we add things to the API, it won't be, right?

Comment (paprikati) [reply internal/provider/incident_user_data_source.go:29]: Like, we can always add these roles later if someone needs them, but all we're doing by including them is making it harder for us to change the representation of the roles later, I think (e.g. changing role slugs)

File: internal/provider/incident_user_data_source.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+func (i *IncidentUserDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_user"
+}
+
+func (i *IncidentUserDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var data IncidentUserDataSourceModel
+	req.Config.Get(ctx, &data)
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:70]: you've got this twice now!

File: internal/provider/incident_user_data_source.go [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+	ID          types.String `tfsdk:"id" json:"id"`
+	Name        types.String `tfsdk:"name" json:"name"`
+	SlackUserID types.String `tfsdk:"slack_user_id" json:"slack_user_id"`
+}
+
+type IncidentUserRequest struct {
+	ID types.String `tfsdk:"id"`
+}
+
+type RBACRole struct {
```
Comment (paprikati) [inline internal/provider/incident_user_data_source.go:38]: ✂️

File: docs/data-sources/user.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+## Schema
+
+### Optional
+
+- `email` (String)
+- `slack_user_id` (String)
+
+### Read-Only
+
+- `base_role` (Attributes) (see [below for nested schema](#nestedatt--base_role))
```
Comment (paprikati) [inline docs/data-sources/user.md:28]: need to regen

**Review by paprikati — APPROVED** [review]
once you've removed those couple of bits!

**Pull request merged**